### ISSUE TITLE
Add installer script that offers a user-friendly way to install the Zarf binary

### ISSUE THREAD

**Issue by lucasrod16** [defenseunicorns/zarf#345]
**Is your feature request related to a problem? Please describe.**
The current documented method of using and learning Zarf involves cloning the zarf git repository locally and running `make` commands. This is a great workflow for people who may want to contribute to developing Zarf because they can inspect the Makefile(s) and see the "how" of installing and using the tool and become familiar with zarf's git repository folder structure.

For users who are more interested in purely learning how to use the tool, a more simple installation process may be desired. An installer script is a popular, convenient method for installing tools like Zarf.  This method only requires a user to copy a `curl` command and a few seconds later be able to run `zarf` commands. This effectively does the same thing as clicking a download link on the releases page, except a script can automatically detect what binary the user needs and install it into a location that is ready for it to be executed against. It would reduce the manual steps involved in getting `zarf` commands to "work".

Another benefit of offering an installer script is that it is automation-friendly. It is a very simple, lightweight solution if a user wants to install the Zarf binary in their own scripts, CI/CD pipelines, and container images.

I believe this would improve UX also because it has become a bit of a norm in this space to offer such a script, so users can get straight to work on learning how to use Zarf versus how to install it. 

Below are a few examples of popular tools in the k8s ecosystem that offer installer scripts for users to download and execute with a one-liner `curl` or `wget` command:

Helm:
- https://helm.sh/docs/intro/install/#:~:text=%24%20curl%20%2DfsSL%20%2Do%20get_helm.sh%20https%3A//raw.githubusercontent.com/helm/helm/main/scripts/get%2Dhelm%2D3
- https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3

Kustomize:
- https://kubectl.docs.kubernetes.io/installation/kustomize/binaries/#:~:text=curl%20%2Ds%20%22https%3A//raw.githubusercontent.com/kubernetes%2Dsigs/kustomize/master/hack/install_kustomize.sh%22%20%20%7C%20bash
- https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh

K3s: 
- https://k3s.io/#:~:text=curl%20%2DsfL%20https%3A//get.k3s.io%20%7C%20sh%20%2D
- https://get.k3s.io/

RKE2:
- https://docs.rke2.io/install/quickstart/#:~:text=curl%20%2DsfL%20https%3A//get.rke2.io%20%7C%20sh%20%2D
- https://get.rke2.io/

**Describe the solution you'd like**
Here is a proposed solution based on Helm's install script (linked above). The design and functionality of the script is fully open to discussion.

PR:
- https://github.com/defenseunicorns/zarf/pull/344

**Describe alternatives you've considered**
Not really an alternative, but adding support for installing via package managers (e.g., Homebrew, apt, yum, dnf, etc) would extend the end-user's options and improve UX for people who prefer to manage installs with their favorite package manager. 

**Additional context**
The updated README.md that's part of the PR (linked above) includes an example of how the install script could be used. It also includes some additional help/suggestions for MacOS users to set an alias for their zarf binary so that `zarf` commands work. I believe adding that documentation supports the Zarf theme of "you shouldn't have to be an expert to use Zarf". I believe that the less assumptions made about end-users, the less potential stress for a beginner getting `zarf` to work. 

- https://github.com/defenseunicorns/zarf/blob/install_script/README.md

**Comment by RothAndrew** [comment]
Thanks for writing this up @lucasrod16. I'm gonna tweak it a little to be more high-level, but I'll make sure to keep what you wrote intact.

**Comment by lucasrod16** [comment]
@RothAndrew no problem! Do your thing 😎

**Comment by RothAndrew** [comment]
_Original Title:_ Add installer script that offers a user-friendly way to install the Zarf binary

_Original Description:_

**Is your feature request related to a problem? Please describe.**
The current documented method of using and learning Zarf involves cloning the zarf git repository locally and running `make` commands. This is a great workflow for people who may want to contribute to developing Zarf because they can inspect the Makefile(s) and see the "how" of installing and using the tool and become familiar with zarf's git repository folder structure.

For users who are more interested in purely learning how to use the tool, a more simple installation process may be desired. An installer script is a popular, convenient method for installing tools like Zarf.  This method only requires a user to copy a `curl` command and a few seconds later be able to run `zarf` commands. This effectively does the same thing as clicking a download link on the releases page, except a script can automatically detect what binary the user needs and install it into a location that is ready for it to be executed against. It would reduce the manual steps involved in getting `zarf` commands to "work".

Another benefit of offering an installer script is that it is automation-friendly. It is a very simple, lightweight solution if a user wants to install the Zarf binary in their own scripts, CI/CD pipelines, and container images.

I believe this would improve UX also because it has become a bit of a norm in this space to offer such a script, so users can get straight to work on learning how to use Zarf versus how to install it. 

Below are a few examples of popular tools in the k8s ecosystem that offer installer scripts for users to download and execute with a one-liner `curl` or `wget` command:

Helm:
- https://helm.sh/docs/intro/install/#:~:text=%24%20curl%20%2DfsSL%20%2Do%20get_helm.sh%20https%3A//raw.githubusercontent.com/helm/helm/main/scripts/get%2Dhelm%2D3
- https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3

Kustomize:
- https://kubectl.docs.kubernetes.io/installation/kustomize/binaries/#:~:text=curl%20%2Ds%20%22https%3A//raw.githubusercontent.com/kubernetes%2Dsigs/kustomize/master/hack/install_kustomize.sh%22%20%20%7C%20bash
- https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh

K3s: 
- https://k3s.io/#:~:text=curl%20%2DsfL%20https%3A//get.k3s.io%20%7C%20sh%20%2D
- https://get.k3s.io/

RKE2:
- https://docs.rke2.io/install/quickstart/#:~:text=curl%20%2DsfL%20https%3A//get.rke2.io%20%7C%20sh%20%2D
- https://get.rke2.io/


**Describe the solution you'd like**
Here is a proposed solution based on Helm's install script (linked above). The design and functionality of the script is fully open to discussion.

PR:
- https://github.com/defenseunicorns/zarf/pull/344

**Describe alternatives you've considered**
Not really an alternative, but adding support for installing via package managers (e.g., Homebrew, apt, yum, dnf, etc) would extend the end-user's options and improve UX for people who prefer to manage installs with their favorite package manager. 

**Additional context**
The updated README.md that's part of the PR (linked above) includes an example of how the install script could be used. It also includes some additional help/suggestions for MacOS users to set an alias for their zarf binary so that `zarf` commands work. I believe adding that documentation supports the Zarf theme of "you shouldn't have to be an expert to use Zarf". I believe that the less assumptions made about end-users, the less potential stress for a beginner getting `zarf` to work. 

- https://github.com/defenseunicorns/zarf/blob/install_script/README.md

**Comment by RothAndrew** [comment]
@lucasrod16 take a look please at the updated description. Does it accurately reflect what you're looking for?

**Comment by RothAndrew** [comment]
Related to https://github.com/defenseunicorns/zarf/issues/295 (selfish feature request, since I'm a heavy user of ASDF. Definitely should not be the only option)

**Comment by lucasrod16** [comment]
@RothAndrew looks good. On Acceptance Criteria # 5, I don't think the goal should be to use less than 3 commands. I think eliminating  the need for  specific knowledge currently required to `curl` a binary is what we're after. Specifically, if you're on a Mac, you have to know what chip you have and then pair that up with the right zarf binary to download. That may be assumed knowledge of the end-users, but if somebody *doesn't* know that information, then that is extra time spent figuring out what system they have and what binary they need. Something like a script should do that for them I think.

**Comment by RothAndrew** [comment]
The question posed to the maintainers is: Is the use of a shell install script, similar to the way projects like Helm, Kustomize, K3s, and RKE do it, appropriate for Zarf? Or do we want to explore other options?

Other common patterns that I can think of off the top of my head are:

* OS-specific package managers (`apt`, `yum`, etc)
* Distro-agnostic package managers (`brew`, `asdf`, `gofish`, etc)
* Git-based
* UI-based (flatpak, .dmg, etc)

**Comment by lucasrod16** [comment]
@RothAndrew much appreciated!

**Comment by RothAndrew** [comment]
> I think eliminating the need for specific knowledge currently required to `curl` a binary is what we're after. Specifically, if you're on a Mac, you have to know what chip you have and then pair that up with the right zarf binary to download. That may be assumed knowledge of the end-users, but if somebody _doesn't_ know that information, then that is extra time spent figuring out what system they have and what binary they need. Something like a script should do that for them I think.

What level of experience with this kind of stuff do we expect from our target user base? So far we have been targeting people that identify as "technicans". People with the skills to maintain systems in disconnected environments, but who aren't kubernetes experts. A submariner is what most frequently comes to mind, and is in fact the basis for Zarf's existence. In that case the technicians on a submarine are highly skilled at the systems they work on, primarily Linux-based, primarily os-level ops with little virtualization or containerization, though that is changing rapidly.

That's not to say there aren't other target user bases, or that this isn't a totally valid feature request, just trying to kick off the conversation.

The best code is the code we don't have to write.

**Comment by RothAndrew** [comment]
As another reference point, the docs to install `kubectl` use pretty much everything EXCEPT "download this shell script and run it". They offer instructions to townload it using `curl`, as well as installation using native package management and "other" package management.

https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/

The likely reason is "download this shell script and run it" can be risky. It requires a lot of trust, unless you go in and inspect the shell script before running it, which would require much more technical expertise than figuring out which binary is needed based on OS and processor architecture.

**Comment by lucasrod16** [comment]
> What level of experience with this kind of stuff do we expect from our target user base? So far we have been targeting people that identify as "technicans". People with the skills to maintain systems in disconnected environments, but who aren't kubernetes experts. A submariner is what most frequently comes to mind, and is in fact the basis for Zarf's existence. In that case the technicians on a submarine are highly skilled at the systems they work on, primarily Linux-based, primarily os-level ops with little virtualization or containerization, though that is changing rapidly.
> 
> That's not to say there aren't other target user bases, or that this isn't a totally valid feature request, just trying to kick off the conversation.
> 
> The best code is the code we don't have to write.

I think providing an installer script makes it more "automation-friendly". If  a users wants to install the zarf binary in a Docker image or wants to include it in other scripts, it can be done with a one-liner `curl` command. If a developer wants to setup a pipeline that automatically creates a zarf package and pushes it to an artifact repository, I think easing the installation of the binary in those situations make sense. Saying this makes me wonder if it would be worth building and publishing Zarf container images for the purpose of using it in things like GitHub Actions and GitLab CI to automate Zarf package creation? I know that may be a bit out of the scope of this issue, but it just popped in my head. Thoughts?

**Comment by RothAndrew** [comment]
I agree with you on it being more automation-friendly. My issues are in other areas.

- I want to make sure we are comfortable enough with the solution that we are willing to support it long-term. It's a fairly significant amount of bash scripting, though to its credit it looks to be mostly copy/pasted from other popular project that do this.
- There is a security risk introduced here by asking someone to download and install a bash script. It isn't a very big one, but it is one that people tend to get jumpy about.

Both of these points are not very big issues, as long as we've documented the reasons why we are making certain decisions, which is what we're doing here. IMO this kind of thing doesn't need an ADR.

Personally:

- I'm okay with the security risk being introduced, as long as we also retain the documentation saying how one can download and install Zarf without needing to use the script. In the long term we should consider providing Zarf in a package manager that is set up to verify cryptographically signed binaries. [Docker's installation docs](https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository) are a good example for this, as they involve downloading Docker's official GPG key and telling the user's package manager to verify the download against that GPG key.

- I'm less okay with supporting the bash script long term. I just don't really think it is necessary. In your examples of creating a docker image it's ~3 lines instead of 1, which isn't a big deal.

What I would be waaay more interested in is supporting a version manager like [ASDF](https://asdf-vm.com) since it would make switching versions really easy for developers of Zarf.

**Comment by jeff-mccoy** [comment]
Tl;Dr; we should probably jump on a call for this.  

This is great to be thinking about the zarf prereq UX, but I think we lack data around any specific issues with the current process and don't see anything in this issue that shows why it is painful now. 

Is this issues saying we "want a more friendly way to download 2 files from your web browser"? If so, I think we're going after the wrong goals here. 

If a user isn't aware that they have intel or apple silicon they aren't going to be a good candidate for using zarf.  You would have to be exceedingly unknowledgeable about tech to not know this. Zarf on anything but Linux requires a cluster which assumes you deployed something via a cloud provider or installer docker + some local distro. All of those things are more technical and complex than zarf itself. 

I do fully agree that make files should not be part of user-facing things at all as they are a poor UX.

I have a lot of concerns around bash scripts for this, that makes assumptions too and I don't want to maintain it to be honest. I'd rather pursue getting zarf into package mgmt systems or use some go release tool to handle updates.  

Lastly, there's a current pr out that will put zarf on docker hub with a signed image that we will put through Iron Bank.

**Comment by lucasrod16** [comment]
@jeff-mccoy @RothAndrew the ultimate idea behind proposing the script was to extend the amount of options users have to install zarf; not saying it's the ideal or best method by any means. Not wanting to maintain the script makes perfect sense and is reason enough not to do it IMO. It sounds like providing install options with package managers is something we can all agree on. Appreciate the conversation!

**Comment by jeff-mccoy** [comment]
Thanks, we would love some help exploring one of those package manager options 😉.
